### ISSUE TITLE
In rbx-2.5.8 String#freeze doesn't act in the same way of ruby-2.0.0

### ISSUE THREAD

**Issue by RoxasShadow** [rubinius/rubinius#3520]
> ruby 2.2.0p0 (2014-12-25 revision 49005) [x86_64-darwin14]

```ruby
2.2.0 :011 > 'a'.freeze.object_id == 'a'.freeze.object_id
 => true 
```

> rubinius 2.5.8 (2.1.0 bef51ae3 2015-10-16 3.5.1 JI) [x86_64-darwin15.0.0]

```ruby
rbx-2.5.8 :015 > 'a'.freeze.object_id == 'a'.freeze.object_id
 => false 
```

If I'm not wrong about the behaviour of rubinius (once you create a frozen String, all the next equal ones will be referenced to the frozen one instead to create a new object) I think that this is a pity considering, for example, the effort that Rails' guys are putting in to reduce the number of object created during the lifetime of a rails application.

**Comment by jemc** [comment]
The difference here is that this behavior of `freeze` called on String literals is a syntax hack in MRI (and in JRuby, because [originally proposed](https://bugs.ruby-lang.org/issues/8992) by them).  In that link, the proposal was to replace the `"my string"f` syntax with `"my string".freeze`, but it still works like syntax rather than working like any other method call, which is one reason I don't think it's a particularly good idea to depend on it as an implementation detail.

That is, with this syntax/compiler hack, `freeze` only works this way when called on a String literal and not a variable reference holding a String:
```ruby
MRI 2.2.2p95 > (x = "a"; x).freeze.object_id == (y = "a"; y).freeze.object_id
```

Since this is a syntax/compiler hack, this is not a matter of changing how `String#freeze` works, but adding an AST transform to the compiler.

**Comment by RoxasShadow** [comment]
I see, so it's something trickier than what expected.

> Also, we'd have to see how MRI and JRuby behave when the String#freeze method is overridden.

I haven't JRuby installed on my machine so I can speak only for MRI: the trick doesn't seem working anymore after `String#freeze` has been overrode ([gist](https://gist.github.com/RoxasShadow/fbc4da5e3809a87497e9)).

**Comment by jemc** [comment]
You can even define `freeze` to call `super` here, and it has the same effect, further demonstrating the "magic" involved here:
```ruby
[1] pry(main)> class String
[1] pry(main)*   def freeze
[1] pry(main)*     super
[1] pry(main)*   end  
[1] pry(main)* end  
=> :freeze
[2] pry(main)> 'a'.freeze.object_id == 'a'.freeze.object_id
=> false
```

**Comment by headius** [comment]
The mixin behavior @jemc shows probably qualifies as a bug. I would imagine MRI is only checking if String#freeze has been replaced, rather than checking if "string".freeze would call a different method.

**Comment by jemc** [comment]
@headius - well, in this case, it is indeed calling a different method - one that happens to call the original within.  What if I had defined it as:
```ruby
module StringMixin
  def freeze
    some_side_effect!
    super
  end
end
```
Should the of the compiler at the `.freeze` call site have to know the difference between that definition and one that only calls `super`?  It's a slippery slope of static analysis.

I'd argue that this behavior is a reasonable optimization for a Ruby implementation to make (in cases where the compiler can be sure that behavior is still correct), but the problems come up when people start looking at an optimization (the reduced object allocations) like a feature of Ruby (that the `object_id`s are the same).  Calling the mixin behavior I showed a bug is part of this problematic mentality.

Sadly, I think this is why the change from `"my string"f` to `"my string".freeze` for frozen string literals was a mistake.  Even if it was just meant to be an optimization, it does have user-visible effects (identity-equivalence), and due to the aggressive introspection in Ruby culture, user-visible effects have a way of becoming part of business logic assumptions even when it is arguably bad practice to do so.  The fact that we have user-visible effects here is what should be an implementation detail look like a bug.  And unfortunately, looking at it like a bug in this case pushes implementations more toward complexity and static analysis (or more ugly hacks).  The `"my string"f` syntax avoids these problems by being completely unambiguous due to not masquerading as a method call.

**Comment by headius** [comment]
"my string"f was also not intended to provide identity; it was always meant to just be a hint to the compiler that this string would be frozen-on-creation. An implementation can implement either "my string"f or "my string".freeze with or without identical objects at each site. MRI and JRuby currently implement it that way, but it's not part of the specification...in fact I mention that it is an open question, and I explicitly point out that implementing this behavior with *no* identity and *no* optimization is 100% valid.

In a later comment, I also recommend that identity *not* be relied upon: https://bugs.ruby-lang.org/issues/8992#note-12

...because the purpose of "my string".freeze was to keep code backward-compatible but allow implementations to optimize pre-frozen strings.

Other discussions in the same issue dropped the idea of having String#freeze always use the internal identity store, making it clear that String#freeze was NOT guaranteed to return identical objects.

And the mixin behavior is an implementation bug in MRI because I believe ko1 wanted to make sure any method table changes that would re-route "my string".freeze would cancel out the compiler optimization.

The bottom line here is that @RoxasShadow should not be relying on "my string".freeze to return the same object every time, and there's no real guarantee of this behavior in *any* implementation. If Rubinius chooses to implement the compiler optimization, there is precedent, but as far as specification goes this improvement is a "may", not a "must".

**Comment by jemc** [comment]
> And the mixin behavior is an implementation bug in MRI because I believe ko1 wanted to make sure any method table changes that would re-route "my string".freeze would cancel out the compiler optimization.

Right, and my mixin example shows it routing through a custom overriden method (that happens to call `super`), so by your explanation and my thinking, it makes sense that the optimization would be invalidated as you describe.  So, not a bug, right?

As for the rest of comment, I'm glad we agree that this should be seen as an implementation detail rather than a specified behavior, and hearing that directly from you (the original instigator of this change in the MRI issue tracker) is helpful for solidifying my assumption.

With that in mind, I'm inclined to close this issue ticket, as the `object_id` is not specified in this case.

**Comment by RoxasShadow** [comment]
Considering what's been said, I agree, this issue can be closed. Thanks!

**Comment by headius** [comment]
@jemc Sorry, I read your comment wrong; I thought that the mixed-in case was still doing the optimization. Since it does not, I agree there's no bug there. Only the case where "my string".freeze calls the standard, built-in String#freeze should the optimization apply.

**Comment by jemc** [comment]
@headius :+1:
